### ISSUE TITLE
CORDEX CMOR tables missing valid_max and valid_min attributes, leading to incorrect preprocessor area operations

### ISSUE THREAD

**Issue by thomascrocker** [ESMValGroup/ESMValCore#1141]
I recently developed a recipe involving a region extraction:
```
start_longitude: -5
end_longitude: 30
start_latitude: 42
end_latitude: 52
```
But, when i plotted the data, it was clear that the region extraction wasn't working correctly on the western boundary.
![MicrosoftTeams-image](https://user-images.githubusercontent.com/17336982/119163128-05f61f80-ba53-11eb-9584-324cf8937bf8.png)

After a bunch of digging, I realised that the issue was that the longitude coord of the original data was defined in the range `[-180, 180]` and the CMOR checker / fixer was not correcting it to `[0, 360]`. The region extraction functions in the preprocessor all assume longitude lives in the range `[0, 360]`, and thus the cropping out of the points beyond 5W was not happening.
The reason for the CMOR checker not fixing the coord was because the default CMOR tables for CORDEX included with ESMValTool don't include the `valid_min` and `valid_max` attributes on the longitude coords, and so all the checks and fixes in https://github.com/ESMValGroup/ESMValCore/blob/b2f7ffc3232f174dd5ebc50ad20b4a02d3517c2c/esmvalcore/cmor/check.py#L617 were skipped.

I've created a branch with a fix for this (by updating the tables). See #1142 

Below is an example of a file from ESGF with the offending longitude data in question. It is not the only CORDEX model to have this issue though. Which does raise the worrying prospect that any recipes people have written to use CORDEX to do for example, area statistics in a region that has parts west of the meridian, have potentially the wrong results for some models with coordinates like this without realising. I only spotted the issue because I produced spatial maps of my data, if I had only been calculating area averages (or other statistics) it wouldn't have been at all obvious that there was any problem.
http://cordexesg.dmi.dk/thredds/fileServer/cordex_general/cordex/output/EUR-11/DMI/MOHC-HadGEM2-ES/historical/r1i1p1/DMI-HIRHAM5/v2/mon/tas/v20190512/tas_EUR-11_MOHC-HadGEM2-ES_historical_r1i1p1_DMI-HIRHAM5_v2_mon_199101-200012.nc

**Comment by zklaus** [comment]
I agree that this is an important issue, but I don't think the solution is to force all longitudes to be in [0, 360]. This is appropriate for global data on regular grids since in that case, it's just a matter of agreeing on where to put the seam.

But for global models with grids that are not aligned with longitude and latitude, and for regional models that have rotated or projected grids, this is less clear.  Indeed, for [the official CMIP6 guidance (Section "Requirements for coordinate variables", Point 9)](https://goo.gl/neswPr) is

> When the coordinate variables for a horizontal grid are not longitude and latitude, it is required that the true latitude and longitude coordinates be supplied via the “coordinates” attribute. [...] it is recommended that when possible, longitudes increase monotonically and be confined to the range -180 to 360 (with their span limited to a single cycle)[...]

in other words, monotonicity is considered more important than limiting this to [0, 360], which makes sense since many calculations become difficult otherwise.

[The CORDEX archive specifications](http://is-enes-data.github.io/cordex_archive_specifications.pdf) don't make an explicit statement about it, but CORDEX uses a lot of negative longitudes, which again makes sense since for regional models, forcing a limit on the longitudes introduces a seam that isn't even there.

So what to do instead? I think we should find out which preprocessors are problematic and assess whether it's feasible to fix them.

**Comment by thomascrocker** [comment]
I can see what you mean, as you say, with a regional model, there is no need for a seam, anywhere, so if introducing one artificially can be avoided it probably would help avoid other potential problems elsewhere.

So I've just done a search for `360` over the repo, and looked at what is thrown up. The main culprit is, unsurprisingly, is `_area.py` specifically https://github.com/ESMValGroup/ESMValCore/blob/fee5d541dc871efe72ab972a6a52195c452d16eb/esmvalcore/preprocessor/_area.py#L26 that is the cause of my problems.

The assumption is also used in https://github.com/ESMValGroup/ESMValCore/blob/fee5d541dc871efe72ab972a6a52195c452d16eb/esmvalcore/preprocessor/_area.py#L353 and https://github.com/ESMValGroup/ESMValCore/blob/fee5d541dc871efe72ab972a6a52195c452d16eb/esmvalcore/preprocessor/_area.py#L302 which are related to region extraction using shapefiles which isn't supported yet for irregular grids, although I guess a regional model on a non rotated pole, e.g one run near the equator would be able to do this sort of extraction, but potentially give incorrect results.

The other place `360` appears prominently is: 
https://github.com/ESMValGroup/ESMValCore/blob/fee5d541dc871efe72ab972a6a52195c452d16eb/esmvalcore/preprocessor/_mask.py#L270
Which is also related to use of shapefiles, this time for masking. 

There are a bunch of other tests etc. as well which would possibly need tweaking, but I think what I mentioned above is the only proper code.

My inkling is that the `extract_region` method could probably be fixed by using an approach that queried the `circular` property of the longitude coordinate. I _think_ that convention is this should always be `True` for GCM data and false for RCM data. If I'm right about that then a check could be introduced to apply the logic in lines 70 to 84 for GCM data with circular coordinates, and a more straightforward comparison for RCM data that should be non circular and strictly monotonically increasing. But can we assume that is _always_ the case for RCM data?

**Comment by zklaus** [comment]
Thanks for investigating this, Detective @thomascrocker!
I think we should assume always monotonic longitudes; of course, in longitude direction, i.e. when looking at neighboring rows of different latitudes, longitudes may well decrease.

I was also wondering a bit more about GCMs. It's true that the placement of the seam is basically just a choice, but this choice is arguably made by the limits of the 2d matrix that contains the data, right? So maybe the best thing really is just to treat GCM and RCM in exactly the same way and figure out how to change the preprocessors.

**Comment by zklaus** [comment]
@thomascrocker, have a look at #245 if you like. I think we can tweak that so that it solves this problem.

**Comment by thomascrocker** [comment]
Thanks #245 does look promising. Unfortunately I have a couple of other things on my plate at the moment that take priority over this but I should hopefully be able to try and work on this again in a few weeks.

**Comment by zklaus** [comment]
No worries, we'll drive #245 forward. Note that the feature freeze for the core for the 2.3.0 release is June 7; anything not in by then will have to wait until the next release in October (c.f. https://docs.esmvaltool.org/en/latest/community/release_strategy.html).

**Comment by thomascrocker** [comment]
Thanks @zklaus, just tested and the additions to #245 resolve this issue.
